route/src/route_generator_worker.cpp

```diff
@@ -438,5 +438,98 @@ namespace route {
         return true;
     }
 
+bool RouteGeneratorWorker::crosstrack_error_check(const geometry_msgs::PoseStampedConstPtr& msg, lanelet::ConstLanelet current, carma_wm::TrackPos llt_track)
+{
+  auto route = lanelet::ConstLanelets();
+  for(auto id : route_msg_.route_path_lanelet_ids)
+  {
+    auto ll = world_model_->getMap()->laneletLayer.get(id);
+    route.push_back(ll);
+  }
+
+  bool out_of_llt_bounds = false;
+
+    lanelet::BasicPoint2d position;
+
+    position.x()= msg->pose.position.x;
+    position.y()= msg->pose.position.y;
+
+
+    ROS_DEBUG_STREAM("Distance1: "<< boost::geometry::distance(position, current.polygon2d())<<" Crosstrack: "<< cross_track_dist );
+    ROS_DEBUG_STREAM("Distance2: "<< boost::geometry::distance(position, current.polygon2d())<<" Downtrack: "<< llt_track.downtrack);
+
+    
+    if (boost::geometry::distance(position, current.polygon2d()) > cross_track_dist) //Evaluate lanelet crosstrack distance from vehicle
+        {
+            cte_count_++;
+
+            if(cte_count_ > 4) //If the distance exceeds the crosstrack distance a certain number of times, report that the route has been departed
+                return true;
```

Comment:
1. We should have a counter that counts up - 1 variable. And another variable cte_count_ (consider renaming to cte_max_count_) which is the maximum allowed count to be reached. So we are actually setting that 4 with cte_count_. This applies to the other counter as well
2.  Also just realized we should only return true only if both counters are above 4. Because what if this "crosstrack" logic returns true, but the "downtrack" logic would have determined the car is still within bounds, and we returned true too early.

Reply:
@lewisid here